lib/catalog.ex

```diff
@@ -0,0 +1,112 @@
+defmodule Catalog do
+  use GenServer
+
+  alias Catalog.{Bucket, Document}
+
+  @type t :: %__MODULE__{
+          ets: :ets.tid(),
+          loaded?: boolean()
+        }
+
+  @enforce_keys [:ets, :loaded?]
+
+  defstruct ets: nil,
+            loaded?: false
+
+  # Client
+
+  @spec start_link(keyword()) :: GenServer.on_start()
+  def start_link(opts) do
+    name = Keyword.get(opts, :name, __MODULE__)
+
+    GenServer.start_link(__MODULE__, opts, name: name)
+  end
+
+  @spec lookup(String.t() | nil, String.t() | nil, Date.t() | nil) ::
+          {:ok, [Document.t()]} | {:error, String.t()}
+  @spec lookup(GenServer.server(), String.t() | nil, String.t() | nil, Date.t() | nil) ::
+          {:ok, [Document.t()]} | {:error, String.t()}
+  def lookup(pid \\ __MODULE__, last_name, first_name, date_of_birth)
+
+  def lookup(_, nil, nil, nil),
+    do: {:error, "Bad request. You must include at least one field to search on."}
+
+  def lookup(pid, last_name, first_name, date_of_birth) do
+    GenServer.call(pid, {:lookup, last_name, first_name, date_of_birth})
+  end
+
+  # Server
+
+  @impl GenServer
+  @spec init(keyword()) :: {:ok, t(), {:continue, {:load_documents, keyword()}}}
+  def init(opts) do
+    ets = :ets.new(__MODULE__, [:set, :protected, {:read_concurrency, true}])
```

Comment:
Since only this GenServer is reading from the table, you don't need read concurrency.

Reply:
Thanks, that's good to understand.